server/src/main/java/org/opensearch/search/aggregations/bucket/FastFilterRewriteHelper.java

```diff
@@ -354,86 +423,118 @@ public DateFieldMapper.DateFieldType getFieldType() {
             assert fieldType instanceof DateFieldMapper.DateFieldType;
             return (DateFieldMapper.DateFieldType) fieldType;
         }
-    }
 
-    public static boolean isCompositeAggRewriteable(CompositeValuesSourceConfig[] sourceConfigs) {
-        return sourceConfigs.length == 1 && sourceConfigs[0].valuesSource() instanceof RoundingValuesSource;
-    }
+        @Override
+        public DebugInfo tryFastFilterAggregation(
+            PointValues values,
+            Ranges ranges,
+            BiConsumer<Long, Long> incrementDocCount,
+            Function<Object, Long> bucketOrd
+        ) throws IOException {
+            int size = Integer.MAX_VALUE;
+            if (this instanceof CompositeAggregator.CompositeAggregationType) {
+                size = ((CompositeAggregator.CompositeAggregationType) this).getSize();
+            }
+
+            DateFieldMapper.DateFieldType fieldType = getFieldType();
+            BiConsumer<Integer, Integer> incrementFunc = (activeIndex, docCount) -> {
+                long rangeStart = LongPoint.decodeDimension(ranges.lowers[activeIndex], 0);
+                rangeStart = fieldType.convertNanosToMillis(rangeStart);
+                long ord = getBucketOrd(bucketOrd.apply(rangeStart));
+                incrementDocCount.accept(ord, (long) docCount);
+            };
 
-    public static long getBucketOrd(long bucketOrd) {
-        if (bucketOrd < 0) { // already seen
-            bucketOrd = -1 - bucketOrd;
+            return multiRangesTraverse(values.getPointTree(), ranges, incrementFunc, size);
         }
 
-        return bucketOrd;
+        private static long getBucketOrd(long bucketOrd) {
+            if (bucketOrd < 0) { // already seen
+                bucketOrd = -1 - bucketOrd;
+            }
+
+            return bucketOrd;
+        }
     }
 
     /**
-     * Try to get the bucket doc counts for the date histogram aggregation
-     * <p>
-     * Usage: invoked at segment level — in getLeafCollector of aggregator
-     *
-     * @param incrementDocCount takes in the bucket key value and the bucket count
+     * For range aggregation
      */
-    public static boolean tryFastFilterAggregation(
-        final LeafReaderContext ctx,
-        FastFilterContext fastFilterContext,
-        final BiConsumer<Long, Integer> incrementDocCount
-    ) throws IOException {
-        fastFilterContext.segments++;
-        if (!fastFilterContext.rewriteable) {
-            return false;
-        }
+    public static class RangeAggregationType implements AggregationType {
```

Comment:
can we package these agg types up and make them package-private?  Its a bit confusing with some living here and others not, ie `CompositeAggregationType`

Reply:
The idea of having AggregationType is to provide RewriteHelper access to any fields in different Aggregators. Only 2 places we would see this, the common logic is still in RewriteHelper, the logic for accessing Aggregator fields are in Aggregator.

Reply:
We should move all the concrete implementations of AggregationType outside of this class. `FastFilterRewriteContext` should be completely oblivious of the concrete implementation that should be owned by the specific Aggregators. Also, does it make sense to rename the `AggregationType` interface as `OptimizableAggregation` or something?

Reply:
Collected idea in the task